MSBuild package versions for dotnet Source Build

dominoFire opened the pull request:
## Bug

Fixes: 

- https://github.com/dotnet/source-build/issues/890
- Partially fixes https://github.com/nuget/home/issues/8139

Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

- Upgrades MSBuild packages to 16.x 
- Adds new project import for source build needs

## Testing/Validation

Tests Added: No
Reason for not adding tests: 
Validation:  Run on CI

Commit:
Enabling source build (WiP)

- Applied patches
- Separated MSBuild package versions variables
- Fails on netcoreapp2.1 support

Commit:
Renamed MSBuild prefix for MicrosoftBuild in MSBuild Tasks projects

Commit:
Unnaplied patches

Commit:
Removed hard-coded Microsoft.Build.Framework package version on NuGetSdkResolver

Commit:
Comments

nkolev92 commented on build/config.props:
```diff
@@ -31,9 +31,16 @@
     <FluentAssertionsVersion>4.19.4</FluentAssertionsVersion>
-    <MicrosoftBuildPackageVersion>15.1.1012</MicrosoftBuildPackageVersion>
+
+    <MicrosoftBuildFrameworkPackageVersion>16.0.461</MicrosoftBuildFrameworkPackageVersion>
```
Why split up the properties if they are going to be the same? 
In fact we should strive to keep them the same, so I think we should prefer having a single variable. 

Any mixing and matching from packages build from the same repo is a likely untested matrix. While semver's guidance would say that's perfectly fine (and in most cases it is), we should still not complicate our lives with potentially having mismatching versions :)

nkolev92 replied:
Furthermore this will minimize your changeset.

rainersigwald replied:
MSBuild's version aren't semver. Please don't mix them :)

dominoFire replied:
Ok, I will leave the MSBuild package versions in separated properties.

nkolev92 replied:
I think our feedback was the exact opposite. 
Change it back to only 1 property. 
Rainer is suggesting to never have different versions of the MSBuild packages throughout our repo.

nkolev92 replied:
@rainersigwald 

If/when you ship patches for MSBUild, do you build and release all the packages in repo? 

For example;
NuGet's patching model is release a new version of everything. 
ASP.NET's patching model is do the best to release a new version of only the packages changing and everything above them (I think).

Which one do you follow?

rainersigwald replied:
We release a new version of everything.

dominoFire replied:
Discussed on chat/mail. We can have the same MicrosoftBuildPackageVersion property and dotnet source-build should be able to override it. We don't need a property for each MSBuild package version.

rainersigwald reviewed (commented):
From the MSBuild perspective, this looks fine, as long as you don't need to run on 15.x any more. Is that the case?

rrelyea commented on src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Microsoft.Build.NuGetSdkResolver.csproj:
```diff
@@ -20,7 +20,7 @@
   <ItemGroup>
     <ProjectReference Include="$(NuGetCoreSrcDirectory)NuGet.Commands\NuGet.Commands.csproj" />
-    <PackageReference Include="Microsoft.Build.Framework" Version="15.6.85" ExcludeAssets="runtime" />
+    <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildFrameworkPackageVersion)" ExcludeAssets="runtime" />
```
1) you are missing the import in this file of the dotnetpackageversionpropspath
2) You need to ensure that this property has a default value when "source build" properties are not being injected into this build. likely change it to match the same version as the rest of NuGet.

(Need review and signoff from msbuild folks on this part of the change)

dominoFire replied:
This project imports the `commons.props` file, which imports the `config.props` file, that has the `DotnetPackageVersionPropsPath` project import: 

https://github.com/NuGet/NuGet.Client/blob/18580becd553aaa61b7dab4d2b7a3779c4c0b0a2/src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Microsoft.Build.NuGetSdkResolver.csproj#L2

The default value is provided in the `config.props` file, modified in this PR.

rrelyea commented on test/NuGet.Core.Tests/Microsoft.Build.NuGetSdkResolver.Tests/Microsoft.Build.NuGetSdkResolver.Test.csproj:
```diff
@@ -11,7 +11,7 @@
   <ItemGroup>
     <ProjectReference Include="$(NuGetCoreSrcDirectory)Microsoft.Build.NuGetSdkResolver\Microsoft.Build.NuGetSdkResolver.csproj" />
-    <PackageReference Include="Microsoft.Build.Framework" Version="15.6.85" />
+    <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildFrameworkPackageVersion)" />
```
1) you are missing the import in this file of the dotnetpackageversionpropspath
2) You need to ensure that this property has a default value when "source build" properties are not being injected into this build. likely change it to match the same version as the rest of NuGet.

(Need review and signoff from msbuild folks on this part of the change)

dominoFire replied:
Same case as above: This project imports the `commons.test.props` file:

https://github.com/NuGet/NuGet.Client/blob/18580becd553aaa61b7dab4d2b7a3779c4c0b0a2/test/NuGet.Core.Tests/Microsoft.Build.NuGetSdkResolver.Tests/Microsoft.Build.NuGetSdkResolver.Test.csproj#L2

which imports the `config.props` file:

https://github.com/NuGet/NuGet.Client/blob/18580becd553aaa61b7dab4d2b7a3779c4c0b0a2/build/common.test.props#L9

that has the `DotnetPackageVersionPropsPath` project import: 

https://github.com/NuGet/NuGet.Client/blob/18580becd553aaa61b7dab4d2b7a3779c4c0b0a2/build/config.props#L35


The default value is provided in the `config.props` file, modified in this PR.

dominoFire commented:
> From the MSBuild perspective, this looks fine, as long as you don't need to run on 15.x any more. Is that the case?

@rainersigwald In this dev branch, we don't build/run on 15.x versions. Sometimes, we do servicing in release-4.x branches

nkolev92 commented on build/config.props:
```diff
@@ -31,9 +31,16 @@
   </PropertyGroup>
 
+  <!-- Special project import for dotnet source build -->
+  <Import Project="$(DotNetPackageVersionPropsPath)" Condition="'$(DotNetPackageVersionPropsPath)' != ''" />
```
Can you elaborate more on why we are doing this.

dominoFire replied:
That is the implementation of the Repo API for dotnet source-build capabilities.
In that import, the dotnet source-build injects a props file with source-builded package versions available at build time.

Commit:
Addressing feedback on

- Added single default value for MSBuild packages
- Prioritizing source-build props project import
- Added comments
- Added warning message for MSBuild package version mismatch

Commit:
Conditional MicrosoftBuildPackageVersion property definition

Commit:
Minimizing changeset

Commit:
More min-changeset

dominoFire commented:
Failing tests:

Run 6264308 - NuGet.Client Apex Tests On Windows /
Tampered_InstallFromPMCForPC_FailAsync (x3)
Documented here: NuGet/Home#8146

Run 6264308 - NuGet.Client Apex Tests On Windows /
UninstallPackageFromUI
Documented here: NuGet/Home#8097

Run 6264224 - NuGet.Client Tests On Mac
RestoreNetCore_PackagesLockFile_CustomAssemblyName_DoesNotBreakLockedMode
RestoreNetCore_PackagesLockFile_CustomPackageId_DoesNotBreakLockedMode
RestoreNetCore_PackagesLockFile_LowercaseProjectNameSolutionRestore
Documented here: NuGet/Home#8114

Run 6208776 - NuGet.Client EndToEnd Tests On Windows / 
NetCoreProjectExecuteInitScriptOnlyOnce
Documented here: NuGet/Home#7891
VariablesPassedToInstallScriptsAreValidWithWebSite
Documented here: https://github.com/NuGet/Home/issues/8135

zivkan commented on build/config.props:
```diff
@@ -24,14 +24,23 @@
+
+  See: https://github.com/dotnet/source-build/blob/master/Documentation/auto-dependency-flow/api.md#-pdotnetpackageversionpropspathpath
+  -->
+  <Import Project="$(DotNetPackageVersionPropsPath)" Condition="'$(DotNetPackageVersionPropsPath)' != ''" />
```
if you move the import to after the dependency versions are defined, that should give Source Build the opportunity to override any version (for example xunit, newtonsoft, moq, etc), even if they don't need to right now, while removing the need for us to put a condition on every property?

dominoFire replied:
Right.

The Repo API documentation says that the imported project will only have Properties with the format PacakageId + "PackageVersion". In that case, Only `<MicrosoftBuildPackageVersion>` would be overriden.

I moved at the beginning because the initial implementation I needed to deal with 3 other properties but, at the end, we only required the original `<MicrosoftBuildPackageVersion>` property with an upgraded version.

I will move the import below.

nkolev92 replied:
The good MSbuild practice is define imports at the top.they could accidentally define properties that we define as well, overwrite them and we would get unexpected errors.

Commit:
Moving Repo API project import

Commit:
Comments polishing

nkolev92 reviewed (changes requested):
Imports at the top.

Commit:
Moving Repo API to the top

Commit:
Adding condition on MicrosoftBuildPackageVersion

donnie-msft commented:
Thanks for explaining the changes. Looks like a good improvement.

donnie-msft reviewed (approved):
Thanks for explaining the changes. Looks like a good improvement.

nkolev92 commented:
@dominoFire 
you might want to create a nuget side tracking issue for this. 
It's makes the insertions easier and allows Rob to generate the release notes better.
